Fix eltype example in Methods section of manual

charmasaur opened the pull request:
~The code as written doesn't work for me when passing `eltype(Array{T} where T <: Integer)` (Julia 1.5.4), IIUC because `Array{T} where T <: Integer` is not a subtype of `AbstractArray{Any}` (but it *is* a subtype of `AbstractArray`).

I guess maybe the example isn't meant to be complete, but I got confused when trying to understand why `eltype(Array{T} where T <: Integer)` should have been `Any` from the code, so maybe others would benefit from a little more detail too.

(Also apologies if I'm missing something obvious here; it's my first day with Julia.)~

My reading of the original example was that it should return `Any` for `UnionAll` types, which it doesn't (instead it's just the actual implementation in `Base` that does that). This PR clarifies that, and also removes the next two examples, which throw errors for certain types (for example unions). It replaces the latter with another section explaining that walking the type hierarchy using `supertype` can be problematic.

charmasaur committed:
Fix eltype example in Methods section of manual

The code as written doesn't work for me when passing `eltype(Array{T} where T <: Integer)` (Julia 1.5.4), IIUC because `Array{T} where T <: Integer` is not a subtype of `AbstractArray{Any}` (but it *is* a subtype of `AbstractArray`).

I guess maybe the example isn't meant to be complete, but I got confused when trying to understand why `eltype(Array{T} where T <: Integer)` should have been `Any` from the code, so maybe others would benefit from a little more detail too.

sostock commented:
What do you mean by

> The code as written doesn't work for me

?

Maybe I don’t understand you, but even without the method you added, `Any` is returned, because there is a fallback `eltype(::Type) = Any`.

charmasaur commented:
@sostock thanks for taking a look!

> What do you mean by [...]

That when I copy the `eltype` definition from this docs section into a REPL, and rename it to avoid conflicting with the built-in one (I guess this is an important point I neglected to mention), I get an error when passing `Array{T} where T <: Integer` (since there's no fallback).

> because there is a fallback `eltype(::Type) = Any`.

Ok, I think in definitely missing something obvious. Where's that fallback defined? If you just mean it's defined in the actual `eltype` implementation in Julia, I think it would be more clear to include it explicitly in the example here.

sostock commented:
> If you just mean it's defined in the actual `eltype` implementation in Julia

Yes, that’s what I mean. I still wouldn’t add the change you propose in this PR, I would rather explain in the text below that the `eltype(Array{T} where T <: Integer) == Any` behavior is due to a generic fallback method.

Just for reference, there is an issue about removing the generic fallback: #34041.

charmasaur commented:
OK, but to me it's a little misleading that the preceding text says "Here is the correct code template for returning the element-type `T` of any arbitrary subtype of `AbstractArray`", but that the code template doesn't cover some subtypes of `AbstractArray`. It's also inconsistent with the next two code fragments, which *do* cover all subtypes of all `AbstractArray`.

sostock commented:
> OK, but to me it's a little misleading that the preceding text says "Here is the correct code template for returning the element-type `T` of any arbitrary subtype of `AbstractArray`", but that the code template doesn't cover some subtypes of `AbstractArray`.

That could be reworded. Here’s how I see it: The given template implementation returns the element type for *any subtype of `AbstractArray` that has a well-defined element type*. For subtypes of `AbstractArray` that do not have a well-defined element type, it returns `Any` via a generic fallback. That fallback applies to all types, not just `AbstractArray`. Something like this could be incorporated into the manual.

The phrase “the correct code template” should probably changed as well, since it implies that it is the *only* correct way even though the next two code blocks are also supposedly correct (see below for why they are actually not correct, or at least not equivalent to the first implementation).

> It's also inconsistent with the next two code fragments, which _do_ cover all subtypes of all `AbstractArray`.

I think the next two code blocks are the ones that are actually problematic, not the first one. They are not equivalent to the first one, because they actually error instead of returning `Any` in the case of `Union`s, since `supertype` isn’t defined for those:
```julia
julia> my_eltype(::Type{A}) where {A<:AbstractArray} = eltype(supertype(A))
my_eltype (generic function with 1 method)

julia> my_eltype(Union{Array{Int}, Array{Float64}})
ERROR: MethodError: no method matching supertype(::Type{Union{Array{Float64,N} where N, Array{Int64,N} where N}})
```
I think that these code blocks should be fixed, especially since the generic fallback `eltype(::Type) = Any` presubably exists to make sure that `eltype` never throws an error. Maybe they could just be removed, leaving only the first code block?

fredrikekre commented:
Perhaps it would be better to use a different function name altogether?

charmasaur committed:
Rework eltype example in Methods section

Removed the two other "correct" examples, and replaced them with another warning about a possible mistake (namely for types without supertypes). Also re-worded the correct example, to make a little more clear where the fallback is defined.

charmasaur commented:
> That could be reworded. Here’s how I see it: The given template implementation returns the element type for _any subtype of `AbstractArray` that has a well-defined element type_. For subtypes of `AbstractArray` that do not have a well-defined element type, it returns `Any` via a generic fallback. That fallback applies to all types, not just `AbstractArray`. Something like this could be incorporated into the manual.

That sounds reasonable to me. I've reworked it a bit to try to make clear where the fallbacks happen.

> The phrase “the correct code template” should probably changed as well, since it implies that it is the _only_ correct way even though the next two code blocks are also supposedly correct (see below for why they are actually not correct, or at least not equivalent to the first implementation).
> [...]
> I think the next two code blocks are the ones that are actually problematic, not the first one. They are not equivalent to the first one, because they actually error instead of returning `Any` in the case of `Union`s, since `supertype` isn’t defined for those:
> [...]

Interesting, thanks for the example! I don't know how to fix those examples in a way that makes sense given the context (i.e. without using triangular dispatch), so I've just removed them, and replaced them with another warning about how things can go wrong.

> Perhaps it would be better to use a different function name altogether?

I think the current wording makes the distinction between the example implementation and the `Base` implementation reasonably clear, but I'd be happy to rename it too.
